core/src/main/java/tech/pegasys/eth2signer/core/multikey/DirectoryBackedArtifactSignerProvider.java

```diff
@@ -57,117 +54,73 @@ public DirectoryBackedArtifactSignerProvider(
-    loadedSigner.ifPresent(signer -> artifactSignerCache.put(normaliseIdentifier, signer));
+  public void loadSigners() {
+    final Collection<ArtifactSigner> allSigners =
+        findSigners(this::matchesFileExtension, signerParser::parse);
+    LOG.info(
+        "Loading signers complete. Signers = {}, {}", allSigners.size(), artifactSigners.size());
+    allSigners
+        .parallelStream()
+        .forEach(
+            signer -> artifactSigners.put(normaliseIdentifier(signer.getIdentifier()), signer));
+    LOG.info("Created artifactSigners = {}", artifactSigners.size());
   }
 
   @VisibleForTesting
-  protected LoadingCache<String, ArtifactSigner> getArtifactSignerCache() {
-    return artifactSignerCache;
-  }
-
-  private Optional<ArtifactSigner> loadSignerForIdentifier(final String signerIdentifier) {
-    final Filter<Path> pathFilter = signerIdentifierFilenameFilter(signerIdentifier);
-    final Collection<ArtifactSigner> matchingSigners = findSigners(pathFilter, signerParser::parse);
-    if (matchingSigners.size() > 1) {
-      LOG.error(
-          "Found multiple signing metadata file matches for signer identifier " + signerIdentifier);
-      return Optional.empty();
-    } else if (matchingSigners.isEmpty()) {
-      return Optional.empty();
-    } else {
-      return Optional.of(matchingSigners.iterator().next());
-    }
+  protected Map<String, ArtifactSigner> getArtifactSignerCache() {
+    return artifactSigners;
   }
 
   private <T> Collection<T> findSigners(
-      final DirectoryStream.Filter<? super Path> filter, final Function<Path, T> mapper) {
-    final Collection<T> signers = new HashSet<>();
-
-    try (final DirectoryStream<Path> directoryStream =
-        Files.newDirectoryStream(configsDirectory, filter)) {
-      for (final Path file : directoryStream) {
-        try {
-          signers.add(mapper.apply(file));
-        } catch (final Exception e) {
-          renderException(e, file.getFileName().toString());
-        }
-      }
-      return signers;
-    } catch (final IOException | SecurityException e) {
-      LOG.warn("Error searching for signing metadata files: {}", e.getMessage());
-      return Collections.emptySet();
+      final Predicate<? super Path> filter, final Function<Path, T> mapper) {
+    try (final Stream<Path> fileStream = Files.list(configsDirectory)) {
+      LOG.info("Config files count = {}", fileStream.count());
+    } catch (final IOException e) {
+      LOG.error("Didn't want this one", e);
     }
-  }
 
-  private Filter<Path> signerIdentifierFilenameFilter(final String signerIdentifier) {
-    return entry -> {
-      final String baseName = FilenameUtils.getBaseName(entry.toString());
-      return matchesFileExtension(entry)
-          && baseName.toLowerCase().endsWith(signerIdentifier.toLowerCase());
-    };
+    try (final Stream<Path> fileStream = Files.list(configsDirectory)) {
```

Comment:
you could have above logging statement here instead of running Files.list twice (L98, L105)

Reply:
the above log line has been removed, but I found you couldn't embed it here, as it makes the stream "dirty" and it therefore can't then be used to access the files (as the stream is already deemed "used" by the call to count()).